x/migration/keeper/msg_server_claim_morse_application.go

```diff
@@ -3,16 +3,135 @@ package keeper
 
-	// TODO: Handling the message
-	_ = ctx
+	if err := msg.ValidateBasic(); err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
 
-	return &migrationtypes.MsgClaimMorseApplicationResponse{}, nil
+	shannonAccAddr, err := cosmostypes.AccAddressFromBech32(msg.ShannonDestAddress)
+	// DEV_NOTE: This SHOULD NEVER happen as the shannonDestAddress is validated
+	// in MsgClaimMorseApplication#ValidateBasic().
+	if err != nil {
+		return nil, status.Error(
+			codes.InvalidArgument,
+			cosmoserrors.ErrInvalidAddress.Wrapf(
+				"failed to parse shannon destination address (%s): %s",
+				msg.ShannonDestAddress, err,
+			).Error(),
+		)
+	}
+
+	// Ensure that a MorseClaimableAccount exists for the given morseSrcAddress.
+	morseClaimableAccount, isFound := k.GetMorseClaimableAccount(
+		sdkCtx,
+		msg.MorseSrcAddress,
+	)
+	if !isFound {
+		return nil, status.Error(
+			codes.NotFound,
+			migrationtypes.ErrMorseApplicationClaim.Wrapf(
+				"no morse claimable account exists with address %q",
+				msg.MorseSrcAddress,
+			).Error(),
+		)
+	}
+
+	// Ensure that the given MorseClaimableAccount has not already been claimed.
+	if morseClaimableAccount.IsClaimed() {
+		return nil, status.Error(
+			codes.FailedPrecondition,
+			migrationtypes.ErrMorseApplicationClaim.Wrapf(
+				"morse address %q has already been claimed at height %d by shannon address %q",
+				msg.MorseSrcAddress,
+				morseClaimableAccount.ClaimedAtHeight,
+				morseClaimableAccount.ShannonDestAddress,
+			).Error(),
+		)
+	}
+
+	// Default to the stake amount recorded in the MorseClaimableAccount.
+	if msg.Stake == nil {
+		msg.Stake = &morseClaimableAccount.ApplicationStake
+	}
+
+	// Mint the totalTokens to the shannonDestAddress account balance.
+	// The application stake is subsequently escrowed from the shannonDestAddress account balance.
+	if err = k.MintClaimedMorseTokens(ctx, shannonAccAddr, morseClaimableAccount.TotalTokens()); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// Set ShannonDestAddress & ClaimedAtHeight (claim).
+	morseClaimableAccount.ShannonDestAddress = shannonAccAddr.String()
+	morseClaimableAccount.ClaimedAtHeight = sdkCtx.BlockHeight()
+
+	// Update the MorseClaimableAccount.
+	k.SetMorseClaimableAccount(
+		sdkCtx,
+		morseClaimableAccount,
+	)
+
+	msgStakeApp := apptypes.NewMsgStakeApplication(
+		shannonAccAddr.String(),
+		*msg.Stake,
+		[]*sharedtypes.ApplicationServiceConfig{msg.ServiceConfig},
+	)
+
+	initialAppStake := cosmostypes.NewInt64Coin(volatile.DenomuPOKT, 0)
+	foundApp, isFound := k.appKeeper.GetApplication(ctx, shannonAccAddr.String())
+	if isFound {
+		initialAppStake = *foundApp.Stake
+	}
+
+	app, err := k.appKeeper.StakeApplication(ctx, logger, msgStakeApp)
+	if err != nil {
+		// DEV_NOTE: StakeApplication SHOULD ALWAYS return a gRPC status error.
+		return nil, err
+	}
+
+	// DEV_NOTE: It is safe to use Coin#Sub() here because #StakeApplication() would error otherwise.
```

Comment:
Use `SafeSub()` and handl the errors. This is ONLY true so long as the app staking implementation doesn't change; e.g., supporting down-staking.